libyul/AsmAnalysis.cpp

```diff
@@ -729,6 +733,9 @@ bool AsmAnalyzer::validateInstructions(evmasm::Instruction _instr, SourceLocatio
 		errorForVM(5430_error, "only available for London-compatible");
 	else if (_instr == evmasm::Instruction::BLOBBASEFEE && !m_evmVersion.hasBlobBaseFee())
 		errorForVM(6679_error, "only available for Cancun-compatible");
+	else if (_instr == evmasm::Instruction::BLOBHASH)
+		// TODO: Upgrade this assertion to an error, similar to the ones above, when Cancun becomes the default EVM version.
```

Comment:
It's not a warning, it's an assertion that we'll never run into this in the current code. We won't forget about it - that's the advantage of it being an assert - it will fail in the PR where we change the default EVM version and then we'll know we have to make it an error. IMO this arrangement is slightly better than having a dead error that we can't really trigger and test. It's also a bit clearer (makes it obvious that this can't be triggered) and we also won't forget to remove the error code from the whitelist. Still, it's not *that* much of a difference so I wouldn't bother changing it in your PR. I'd have suggested it there but I only realized we could do this after you merged it. This will be irrelevant anyway once we change the default.

> why doesn't it have a `cancun` check (i.e. `hasBlobHash`)

That's actually a good question. I mean, in practice it does not matter since the only thing it changes is the return value from `validateInstructions()` in case where the opcode exists, but that's [the one case where we ignore that return value](https://github.com/ethereum/solidity/blob/eabe457bf94cd785e0b097b4c03e54572101d29a/libyul/AsmAnalysis.cpp#L327C3-L327C23).

The way the checks are done for other opcodes above, we return `false` when they exist, which seems wrong. @r0qs is correct (it will return `true` instead), but also inconsistent with them. I guess we should fix that by making them consistent and not ignoring the result (in a separate PR), but it's also just a very low priority thing since it won't really affect current behavior in any way.

Reply:
I don't see how having an assert here is better than having a test pinned to `shanghai` that will also fail when we upgrade to `cancun` - compare [blobbasefee test](https://github.com/ethereum/solidity/pull/14755/files#diff-173b4adaa431f4cfc08788f4d23ba19ad95d255388294bda53c9b0d7a6204bd8) with the [basefee test](https://github.com/ethereum/solidity/pull/14755/files#diff-77a7a77f2a2831156ab9f1652dabc5af0858e8ff6739dd3890540ff1d3c5f875); the former will fail due to unmet expectations (dead error becoming live) when we upgrade, and I'd prefer to review updates tests rather than code changes when we do upgrade.

Reply:
I think it's better because it it's clearer. This part of code definitely needs more clarity - it already required too much effort to figure out what's happening there and why. This case is impossible to hit right now and the assert expresses exactly that.

Reply:
> The way the checks are done for other opcodes above, we jump to `else` and return `false` when they do exist, i.e. when the validation is successful, which seems wrong. @r0qs's check is correct (it will return `true` instead), but also inconsistent with that behavior. I guess we should fix that by making them consistent and not ignoring the result (in a separate PR), but it's also just a very low priority cosmetic thing since it won't really affect current behavior in any way.

Indeed. I will move that condition to a separated `if` outside the that chain of checks. But yeah, the idea is exactly that, that it should never reach this point if the instruction is `BLOBHASH` and the current EVM version does not support it.

Reply:
I mean, I think it was better (and less verbose) at the original spot. I'd just do `yulAssert(false)` instead of reporting an error. But well, not like it matters that much.

One thing I think you should adjust though is to put the comment explaining the errors actually above the errors, not above the assert. Otherwise it's likely that it will be removed by mistake along with the assert when we switch the default.

Reply:
Oh right. Sorry, my bad, I got a bit confused by the discussion. I moved it back because I agree that the initial change was better. So hopefully, we can get this merged ;)